File: gnd/src/test/java/com/google/android/gnd/persistence/local/room/converter/ResponseJsonConverterTest.java
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.gnd.persistence.local.room.converter;


import static com.google.common.truth.Truth.assertThat;

import com.google.android.gnd.model.form.Field;
import com.google.android.gnd.model.form.Field.Type;
import com.google.android.gnd.model.observation.DateResponse;
import com.google.android.gnd.model.observation.Response;
import com.google.android.gnd.model.observation.TimeResponse;
import com.google.android.gnd.persistence.local.LocalDatabaseModule;
import com.google.android.gnd.rx.SchedulersModule;
import dagger.hilt.android.testing.HiltAndroidTest;
import dagger.hilt.android.testing.HiltTestApplication;
import dagger.hilt.android.testing.UninstallModules;
import java.util.Date;
import java8.util.Optional;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

public class ResponseJsonConverterTest {

  //Date represented in YYYY-MM-DDTHH:mmZ Format from 1632501600000L milliseconds.
  private static final String DATE_STRING = "2021-09-21T07:00+0000";
  //Date represented in milliseconds for date: 2021-09-24T16:40+0000.
  private static final Date DATE = new Date(1632207600000L);

  @Test
  public void testDateToIsoString() {
    String dateToIsoString = ResponseJsonConverter.dateToIsoString(DATE);
    assertThat(DATE_STRING).isEqualTo(dateToIsoString);
  }

  @Test
  public void testIsoStringToDate() {
    Date stringToDate = ResponseJsonConverter.isoStringToDate(DATE_STRING);
    assertThat(DATE).isEqualTo(stringToDate);
  }

  @Test
  public void testDateToIso_mismatchDate() {
    Date currentDate = new Date();
    Date stringToDate = ResponseJsonConverter.isoStringToDate(DATE_STRING);
    boolean checkMismatchDate = currentDate.equals(stringToDate);
    assertThat(checkMismatchDate).isEqualTo(false);
  }

  @Test
  public void testResponseToObject_dateResponse() {
    Object response = ResponseJsonConverter.toJsonObject(new DateResponse(DATE));
    assertThat(response).isInstanceOf(Object.class);
  }

  @Test
  public void testResponseToObject_timeResponse() {
    Object response = ResponseJsonConverter.toJsonObject(new TimeResponse(DATE));
    assertThat(response).isInstanceOf(Object.class);
  }

  @Test
  public void testObjectToResponse_dateResponse() {
    Object dateObject = ResponseJsonConverter.toJsonObject(new DateResponse(DATE));
    Optional<Response> response = ResponseJsonConverter.toResponse(Field
        .newBuilder()
        .setId("1")
        .setLabel("date")
        .setIndex(0)
        .setRequired(true)
        .setType(Type.DATE)
        .build(), dateObject);
    assertThat(((DateResponse) response.get()).getDate()).isEqualTo(DATE);
  }

  @Test
  public void testObjectToResponse_timeResponse() {
    Object timeObject = ResponseJsonConverter.toJsonObject(new TimeResponse(DATE));
    Optional<Response> response = ResponseJsonConverter.toResponse(Field
        .newBuilder()
        .setId("2")
        .setLabel("time")
        .setIndex(1)
        .setRequired(true)
        .setType(Type.TIME)
        .build(), timeObject);
    assertThat(((TimeResponse) response.get()).getTime()).isEqualTo(DATE);
  }
}

File: ground/src/test/java/com/google/android/ground/persistence/local/room/converter/ResponseJsonConverterTest.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.ground.persistence.local.room.converter

import com.google.android.ground.model.geometry.Coordinates
import com.google.android.ground.model.geometry.LinearRing
import com.google.android.ground.model.geometry.Point
import com.google.android.ground.model.geometry.Polygon
import com.google.android.ground.model.submission.DateResponse
import com.google.android.ground.model.submission.GeometryTaskResponse
import com.google.android.ground.model.submission.MultipleChoiceResponse
import com.google.android.ground.model.submission.NumberResponse
import com.google.android.ground.model.submission.TextResponse
import com.google.android.ground.model.submission.TimeResponse
import com.google.android.ground.model.submission.Value
import com.google.android.ground.model.task.MultipleChoice
import com.google.android.ground.model.task.Option
import com.google.android.ground.model.task.Task
import com.google.common.truth.Truth.assertThat
import com.sharedtest.FakeData
import java.util.Date
import kotlinx.collections.immutable.persistentListOf
import org.json.JSONArray
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.ParameterizedRobolectricTestRunner

@RunWith(ParameterizedRobolectricTestRunner::class)
class ResponseJsonConverterTest(
  private val task: Task,
  private val value: Value,
  private val responseObject: Any
) {

  @Test
  fun testToJsonObject() {
    assertThat(ResponseJsonConverter.toJsonObject(value)).isEqualTo(responseObject)
  }

  @Test
  fun testToResponse() {
    assertThat(ResponseJsonConverter.toResponse(task, responseObject)).isEqualTo(value)
  }

  companion object {
    // Date represented in YYYY-MM-DDTHH:mmZ Format from 1632501600000L milliseconds.
    private const val DATE_STRING = "2021-09-21T07:00+0000"

    // Date represented in milliseconds for date: 2021-09-24T16:40+0000.
    private val DATE = Date(1632207600000L)

    private val multipleChoiceOptions =
      persistentListOf(
        Option("option id 1", "code1", "Option 1"),
        Option("option id 2", "code2", "Option 2"),
      )

    private val singleChoiceResponse =
      MultipleChoiceResponse.fromList(
        MultipleChoice(multipleChoiceOptions, MultipleChoice.Cardinality.SELECT_ONE),
        listOf("option id 1")
      )

    private val singleChoiceResponseJson = JSONArray().apply { put("option id 1") }

    private val multipleChoiceResponse =
      MultipleChoiceResponse.fromList(
        MultipleChoice(multipleChoiceOptions, MultipleChoice.Cardinality.SELECT_MULTIPLE),
        listOf("option id 1", "option id 2")
      )

    private val multipleChoiceResponseJson =
      JSONArray().apply {
        put("option id 1")
        put("option id 2")
      }

    private val pointGeometryTaskResponse =
      GeometryTaskResponse.fromGeometry(Point(Coordinates(10.0, 20.0)))

    private const val pointGeometryTaskResponseString = "HQoFcG9pbnQSFAoSCQAAAAAAACRAEQAAAAAAADRA\n"

    private val polygonGeometryTaskResponse =
      GeometryTaskResponse.fromGeometry(
        Polygon(
          LinearRing(
            listOf(
              Coordinates(10.0, 20.0),
              Coordinates(20.0, 30.0),
              Coordinates(30.0, 40.0),
              Coordinates(10.0, 20.0)
            )
          )
        )
      )

    private const val polygonGeometryTaskResponseString =
      "XQoHcG9seWdvbhJSClAKEgkAAAAAAAAkQBEAAAAAAAA0QAoSCQAAAAAAADRAEQAAAAAAAD5AChIJ\n" +
        "AAAAAAAAPkARAAAAAAAAREAKEgkAAAAAAAAkQBEAAAAAAAA0QA==\n"

    @JvmStatic
    @ParameterizedRobolectricTestRunner.Parameters(name = "task:{0}, data:{1}, jsonObject:{2}")
    fun data() =
      listOf(
        arrayOf(
          FakeData.newTask(type = Task.Type.TEXT),
          TextResponse.fromString("sample text"),
          "sample text"
        ),
        arrayOf(
          FakeData.newTask(type = Task.Type.MULTIPLE_CHOICE),
          singleChoiceResponse,
          singleChoiceResponseJson
        ),
        arrayOf(
          FakeData.newTask(type = Task.Type.MULTIPLE_CHOICE),
          multipleChoiceResponse,
          multipleChoiceResponseJson
        ),
        arrayOf(
          FakeData.newTask(type = Task.Type.NUMBER),
          NumberResponse.fromNumber("12345.0"),
          12345.0
        ),
        arrayOf(FakeData.newTask(type = Task.Type.DATE), DateResponse.fromDate(DATE), DATE_STRING),
        arrayOf(FakeData.newTask(type = Task.Type.TIME), TimeResponse.fromDate(DATE), DATE_STRING),
        arrayOf(
          FakeData.newTask(type = Task.Type.DROP_A_PIN),
          pointGeometryTaskResponse,
          pointGeometryTaskResponseString
        ),
        arrayOf(
          FakeData.newTask(type = Task.Type.DRAW_POLYGON),
          polygonGeometryTaskResponse,
          polygonGeometryTaskResponseString
        ),
      )
  }
}
